R/background.R

```diff
@@ -0,0 +1,253 @@
+
+      tryCatch(
+        wait_until(private$server_is_responsive_and_alive,
+                   timeout = private$timeout,
+                   verbose = private$verbose,
+                   title = "Waiting for server to become responsive"),
+        error = function(e) {
+          process <- private$process
+          private$process <- NULL # ensure always removed on failure
+          porcelain_background_error(e, process, self$log)
+        })
+
+      invisible(self)
+    },
+
+    ##' @description Return the background server status. This will be one of:
+    ##'
+    ##' * `running`: The server is running
+    ##' * `stopped`: The server is stopped
+    ##' * `blocked`: The server is stopped, but something else is running
+    ##'    on the port that we would use
+    ##' * `starting`: The server is starting up (not visible in normal usage)
+    status = function() {
+      is_responsive <- private$server_is_responsive()
+      is_alive <- private$server_is_alive()
+      background_status_string(is_alive, is_responsive)
+    },
+
+    ##' @description Stop a running server. If the server is not running,
+    ##' this has no effect.
+    stop = function() {
+      if (private$server_is_alive()) {
+        if (private$verbose) {
+          message("Stopping server")
+        }
+        private$process$kill()
+        private$process <- NULL
+      }
+      invisible(self)
+    },
+
+    ##' @description Create a url string for the server, interpolating the
+    ##'   (possibly random) port number. You can use this in your tests
+    ##'   like `bg$url("/path")`
+    ##'
+    ##' @param path String representing the absolute path
+    url = function(path) {
+      sprintf("http://localhost:%d%s", self$port, path)
+    },
+
+    ##' @description Run a request to the server, using `httr`. This presents
+    ##'   a similar inteface to the `request` method on the porcelain object.
+    ##'
+    ##' @param method The http method as a string (e.g., `"GET"`), passed
+    ##'   to [httr::VERB] as the `verb` argument
+    ##'
+    ##' @param path String representing the absolute path, passed to `$url()`
+    ##'
+    ##' @param ... Additional arguments passed to `httr::VERB`, such as
+    ##'   `query`, or the body for a `POST` request.
+    request = function(method, path, ...) {
+      httr::VERB(method, self$url(path), ...)
+    }
+  ))
+
+
+## See callr:::default_load_hook for what looks like a part
+## implemented version of an extendable hook; if that support is
+## completed we can use that a little more easily, though there is
+## also no explicit options interface to r_bg either; in any case this
+## is straightforward enough.
+r_bg_with_hook <- function(func, args = list(), stdout = "|", stderr = "|",
+                           user_hook = NULL) {
+  options <- callr::r_process_options(
+    func = func,
+    args = args,
+    stdout = stdout,
+    stderr = stderr,
+    env = callr::rcmd_safe_env())
```

Comment:
Can we have a way to pass an env var to this? In hintr we have a couple of tests which run the real model which we control via
```
withr::with_envvar(c("USE_MOCK_MODEL" = "false"), {
    test_server <- porcelain::porcelain_background$new(
      api, args = list(queue_id = queue_id,
                       results_dir = results_dir))
    test_server$start()
  })
```

This is not working after switching to `porcelain_background` I think because of this hook

Reply:
good find - two options for doing this:

* we add an 'env' argument and add these to the env list
* we inherit from the parent process (which would pick up the withr one there)

We could also do both! What do you think?

Reply:
We might also move hintr to take this as a more conventional argument to the api constructor too?

Reply:
I've propagated `PORCELAIN_VALIDATE` automatically too